### PR TITLE
[Merged by Bors] - feat: lint on (syntactically) duplicate imports

### PR DESCRIPTION
**Pull request by euprunin**
Motivation: PR #16273

### CODE REVIEW

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 1]
```diff
@@ -190,6 +193,7 @@ def parse?_errorContext (line : String) : Option ErrorContext := Id.run do
             some (StyleError.broadImport BroadImports.TacticFolder)
           else
             some (StyleError.broadImport BroadImports.Lake)
+        | "ERR_UIMP" => some (StyleError.duplicateImports)
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/TextBased.lean:196]: Careful, the error code does not match what you had above.

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 2]
```diff
@@ -90,6 +91,7 @@ def StyleError.errorMessage (err : StyleError) : String := match err with
       "In the past, importing 'Lake' in mathlib has led to dramatic slow-downs of the linter (see \
       e.g. mathlib4#13779). Please consider carefully if this import is useful and make sure to \
       benchmark it. If this is fine, feel free to allow this linter."
+  | StyleError.duplicateImports => "Duplicate imports"
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/TextBased.lean:94]: How about including a `name` of a duplicate import in the error message?

Comment (grunweg) [reply Mathlib/Tactic/Linter/TextBased.lean:94]: On the naming side, I'd call the variant `duplicateImport` (singular).
Perhaps it's also nicer to start the error message in lower-case, but as all other variants use upper-case also, that change should be made separately for all variants.

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 3]
```diff
@@ -276,6 +280,30 @@ def adaptationNoteLinter : TextbasedLinter := fun lines ↦ Id.run do
...
+/-- Lint on a collection of input strings if one of the is a duplicate import statement. -/
+def duplicateImportsLinter : TextbasedLinter := fun lines ↦ Id.run do
+  let mut lineNumber := 1
+  let mut errors := Array.mkEmpty 0
+  let mut importStatements := Array.mkEmpty 0
+  for line in lines do
+    if line.startsWith "import " then
+      let lineWithoutComment := (line.splitOn "--")[0]!
+      let mut strippedLine := stripSpaces lineWithoutComment
+        if importStatements.contains strippedLine then
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/TextBased.lean:301]: This should work, but is quadratic in the number of imports. (In practice, that should not matter much.)
A slightly more efficient way would be to sort the imports at the end, and then check if any two consecutive imports are the same (using `List.eraseReps` or similar).

**Review by grunweg — COMMENTED** [review]
Thanks for the PR! While I don't expect this linter to find many *new* issues, it also looks like it should run quickly, so including this as a linter should not hurt.

I have the feeling the implementation can be made more robust, by not manually parsing the list of imports. This is a pre-existing issue; refactoring the `broadImports` linter has been on my list for a while. In the short term, I can think of the following options.
(1) Landing the linter as-is, with my comments addressed (and refactoring in a follow-up).
(2) Extracting the current implementation for parsing the imports into one function, and using that for both linters: this is a small amount of extra work compared to (1), but makes the future refactoring easier to complete.
(3) I'll try to refactor the linters now, and this PR waits on that refactoring. Nicest in the medium-term. I will try, but cannot guarantee I'd get to this before end of September.

What do you think? Do you think you could do (2)? Or would you like to try (3), with some pointers in the right direction?

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 4]
```diff
@@ -5,6 +5,7 @@ Authors: Michael Rothgang
 -/
 
 import Batteries.Data.String.Matcher
+import Lean
```
Comment (joneugster) [inline Mathlib/Tactic/Linter/TextBased.lean:8]:
```suggestion
import Std.Data.HashMap.Basic
```

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 5]
```diff
@@ -53,15 +54,16 @@ deriving BEq
...
+  | Authors
   /-- The bare string "Adaptation note" (or variants thereof): instead, the
   #adaptation_note command should be used. -/
-  | adaptationNote
+  | AdaptationNote
   /-- Lint against "too broad" imports, such as `Mathlib.Tactic` or any module in `Lake`
   (unless carefully measured) -/
-  | broadImport (module : BroadImports)
+  | BroadImport (module : BroadImports)
+  | DuplicateImport (importStatement: String) (alreadyImportedLine: ℕ)
```
Comment (joneugster) [inline Mathlib/Tactic/Linter/TextBased.lean:66]: I don't think this is following the naming convention, is it? My intuition says `StyleError.copyright` should be lower case

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 6]
```diff
@@ -272,7 +278,35 @@ def adaptationNoteLinter : TextbasedLinter := fun lines ↦ Id.run do
...
+  return errors
+
+/-- Strip leading and trailing spaces. -/
+def stripSpaces (s : String) : String := Id.run do
+  let mut stripped := s
+  while stripped.startsWith " " do
+    stripped := stripped.stripPrefix " "
+  while stripped.endsWith " " do
+    stripped := stripped.stripSuffix " "
+  return stripped
```
Comment (joneugster) [inline Mathlib/Tactic/Linter/TextBased.lean:292]: Isn't this `String.trim`?

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 2 above)
Comment (joneugster) [reply Mathlib/Tactic/Linter/TextBased.lean]: I think @grunweg meant the capitalisation of the String messages, not the constructor names!

**Comment by joneugster** [comment]
@grunweg a question, not directly related to getting this PR's scope:

Shouldn't this and the `BroadImport` linter both be syntax based and look at `(← getEnv).header.imports`? 

Or maybe a wrapper of that like [Lean.Environment.importsOf](https://leanprover-community.github.io/mathlib4_docs/ImportGraph/Imports.html#Lean.Environment.importsOf)

(forgot if that still includes the duplicates, but I thought it might)

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 2 above)
Comment (joneugster) [reply Mathlib/Tactic/Linter/TextBased.lean]: So let's just revert this change of yours and stick to 

```
| StyleError.duplicateImports => "Duplicate imports"
```

as you had it originally!

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 5 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/TextBased.lean]: Indeed, I had referred to the capitalisation of the *error messages*, not the constructors. Thanks for correcting this.

**Comment by grunweg** [comment]
PRs are squash-merged in the end, so there is no need to squash intermediate commits.
In fact, it is easier for reviewers if you *don't* rebase your commits once they were reviewed: that makes it easier to see what you changed. (Before your PR gets reviewed, rebasing is fine. I do so quite often when dependent PRs have been merged.)

**Comment by joneugster** [comment]
Im currently just quickly labelling all PRs without a topic label to increase visibility. I'll have a look later

**Comment by joneugster** [comment]
I think this looks very good now, thank you for the PR.

My suggestion to unify this and the broad import linter in a syntax linter can -- if possible -- be a new PR and is certainly out-of-scope here.

maintainer merge

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**